### PR TITLE
[aoti] Add packaging solution

### PR DESCRIPTION
**Pull request by angelayi**
In this PR, I added support for packaging the AOTI generated files into a zipfile, and loading it in python.

`compile_so` takes the path to the package, a device, and a desired so_path location, and compiles package into a .so, and saves to the specified location.
`load_package` takes a path to the package and device, calls _extract_so, and then creates a callable to run the compiled model.

The zipfile generated looks like the following:
```
|- version
|- archive_format
|- data
   |- aotinductor
      |- cbtnafqaqrhvwztv7xudlal4xs6sofxa5oxccyuaqtrt6aozaklx.cubin  # AOTI cuda generated cubin files
      |- cskkqtna23bty2v3aq7g2q37cxrgufehlkuaaolhlgug5zg6fuwe.cpp  # AOTI generated cpp file
      |- cskkqtna23bty2v3aq7g2q37cxrgufehlkuaaolhlgug5zg6fuwe_compile_flags  # Flags for compiling the .o
      |- c6qqtnpgwfi3dv5nb76ai773kt45ezoxfwdmd7q37lvq6fs2tnoi.o  # AOTI saved const.o
      |- cskkqtna23bty2v3aq7g2q37cxrgufehlkuaaolhlgug5zg6fuwe_linker_flags  # Flags for linking the files to form the .so
   |- constants
      |- constants.pt  # Constants saved using torch.save, can be loaded using mmap
```

The workflow is something like:
```
with torch.no_grad():
    ep = torch.export.export(
        model,
        example_inputs,
        dynamic_shapes=dynamic_shapes,
        strict=False,
    )
    gm = ep.module()
    package_path = torch._inductor.aot_compile(
        gm, 
        example_inputs, 
        options= {
              "aot_inductor.output_path": "my_path.pt2",  # or a directory
              "aot_inductor.package": True,
        }
    )
compiled_model = torch._inductor.package.load_package(package_path, device)
return compiled_model
```

I tried turning on loading the weights using mmap by default, but had some trouble with it, so that is just left as a todo

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/package/package.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+from torch._inductor import config, exc
+from torch._utils_internal import get_file_path_2
+from torch.export._tree_utils import reorder_kwargs
+
+
+ARCHIVE_VERSION = 0
+
+with open(
+    get_file_path_2(os.path.dirname(__file__), "pt2_archive_constants.json")
+) as f:
```
Comment (malfet) [inline torch/_inductor/package/package.py:24]: Nit (if one to import `from pathlib import Path)
```suggestion
with (Path(__file__).parent / "pt2_archive_constants.json").open() as f:
```

File: torch/_inductor/package/package.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+AOTINDUCTOR_DIR = PT2_ARCHIVE_CONSTANTS["AOTINDUCTOR_DIR"]
+CONSTANTS_DIR = PT2_ARCHIVE_CONSTANTS["CONSTANTS_DIR"]
+
+
+class PT2ArchiveWriter:
+    def __init__(self, archive_path: str):
+        self.archive_file = zipfile.ZipFile(archive_path, "w")
+        self.write_string("version", str(ARCHIVE_VERSION))
+        self.write_string("archive_format", "pt2")
```
Comment (malfet) [inline torch/_inductor/package/package.py:35]: Shouldn't logic from `__init__` be moved to the enter? (and exit should do `self.archive_file.close`?)

File: torch/_inductor/package/package.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    def write_string(self, name: str, data: Union[bytes, str]) -> None:
+        self.archive_file.writestr(name, data)
+
+    def write_file(self, name: str, file_path: str) -> None:
+        """
+        Copy a file into the archive.
+        name: The destination file inside the archive.
+        file_path: The source file on disk.
+        """
+        assert os.path.isfile(file_path), f"{file_path} is not a valid file path"
```
Comment (malfet) [inline torch/_inductor/package/package.py:52]:
```suggestion
        assert Path(file_path).is_file(), f"{file_path} is not a valid file path"
```

**Review by malfet — APPROVED** [review]

**Comment by angelayi** [comment]
Dependent on https://github.com/pytorch/pytorch/pull/130127/

**Comment by xuhancn** [comment]
> Dependent on #130127

FYI: https://github.com/pytorch/pytorch/issues/124245#issuecomment-2211889758 still not sure if we can land the PR successful.

File: torch/_inductor/package/package.py (hunk 1 above)
Comment (angelayi) [reply torch/_inductor/package/package.py]: this seems to not work with CI ([example](https://github.com/pytorch/pytorch/actions/runs/9896316871/job/27339474170#step:20:766) :/

File: torch/_inductor/package/package.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+import torch._inductor
+import torch.utils._pytree as pytree
+from torch._inductor import config, exc
+from torch._inductor.cpp_builder import BuildOptionsBase, CppBuilder
+from torch.export._tree_utils import reorder_kwargs
+from .build_package import build_package_contents
+from .pt2_archive_constants import AOTINDUCTOR_DIR
+
+
+ARCHIVE_VERSION = 0
```
Comment (malfet) [inline torch/_inductor/package/package.py:21]: This could be part of the constants...

File: torch/_inductor/package/package.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        self.archive_file = zipfile.ZipFile(
+            self.archive_path, "w", compression=zipfile.ZIP_STORED
+        )
+        self.writestr("version", str(ARCHIVE_VERSION))
+        self.writestr("archive_format", "pt2")
+        return self
+
+    def __exit__(self, *args) -> None:  # type: ignore[no-untyped-def]
+        if self.archive_file is not None:
+            self.archive_file.close()
```
Comment (malfet) [inline torch/_inductor/package/package.py:39]:
```suggestion
            self.archive_file.close()
            self.archive_file = None
```

File: torch/_inductor/package/package.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+ARCHIVE_VERSION = 0
+
+
+class PT2ArchiveWriter:
+    def __init__(self, archive_path: str):
+        self.archive_path: str = archive_path
+        self.archive_file: Optional[zipfile.ZipFile] = None
+
+    def __enter__(self) -> "PT2ArchiveWriter":
+        self.archive_file = zipfile.ZipFile(
```
Comment (malfet) [inline torch/_inductor/package/package.py:30]: Nit
```suggestion
        assert self.archive_file is None
        self.archive_file = zipfile.ZipFile(
```

File: torch/_inductor/package/package.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+            self.archive_path, "w", compression=zipfile.ZIP_STORED
+        )
+        self.writestr("version", str(ARCHIVE_VERSION))
+        self.writestr("archive_format", "pt2")
+        return self
+
+    def __exit__(self, *args) -> None:  # type: ignore[no-untyped-def]
+        if self.archive_file is not None:
```
Comment (malfet) [inline torch/_inductor/package/package.py:38]: Should one just assert that it's not None at this point?

File: torch/_inductor/package/package.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+def _run_command_and_check(cmd: str) -> None:
+    cmd = shlex.split(cmd)
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError as e:
+        raise exc.CppCompileError(cmd, e.output) from e
+
+
+def compile_so(aoti_dir: Union[Path, str], aoti_files: List[str], so_path: str) -> str:
```
Comment (malfet) [inline torch/_inductor/package/package.py:98]: This code wouldn't really work when `aoti_dir` is `Path`, please change the signature
```suggestion
def compile_so(aoti_dir: str, aoti_files: List[str], so_path: str) -> str:
```

**Pull request closed without merging**